cfgov/unprocessed/js/modules/util/tree-traversal.js

```diff
@@ -0,0 +1,82 @@
+'use strict';
+
+/**
+ * Perform a backtrack up a Tree to the root.
+ * Given this Tree and starting at (C):
+ *
+ *        R
+ *       / \
+ *      A   B
+ *    / | \
+ *   C  D  E
+ *
+ * Returns (C) -> (A) -> (R).
+ *
+ * @param {TreeNode} node - Node within a tree to traverse from.
+ * @param {Function} callback - Function to call at each node.
+ */
+function backtrack( node, callback ) {
+  callback.call( this, node );
+  var parent = node.parent;
+  if ( parent ) {
+    backtrack( parent, callback );
+  }
+}
+
+/**
+ * Perform an iterative breadth-first search of a Tree.
+ * Given this Tree and starting at (R):
+ *
+ *        R
+ *       / \
+ *      A   B
+ *    / | \
+ *   C  D  E
+ *
+ * Returns (R) -> (A) -> (B) -> (C) -> (D) -> (E).
+ *
+ * @param {TreeNode} node - Node within a tree to traverse from.
+ * @param {Function} callback - Function to call at each node.
+ */
+function bfs( node, callback ) {
+  var queue = [ node ];
```

Comment:
> So why not just loop the children? You could still call the callback for the parent. Could you make the dfs function generic?

Not sure I follow, do you mean utilize the dfs function so it can be called from the bfs?
Order is going to matter. If, while looping over the children, the grandchild (children of the children) are visited before the siblings, it will be a depth-first search, not a breadth-first search. The queue is used to push the grandchildren into a queue that are then visited after all siblings have been visited.